### ISSUE TITLE
make it easier to copy matching results to clipboard

### ISSUE THREAD

**Issue by bbelderbos** [PyBites-Open-Source/search#38]
I often find myself cleaning up the pipes (`|`), e.g.

```
?146 Downloads  $ vid main
                                                        Pybites YouTube Videos
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Title                                                                                 ┃ Url                                         ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│ Pybites Podcast 131 - Write more maintainable Python code, avoid these 15 code smells │ https://www.youtube.com/watch?v=E8K0eS7dHvk │
│ How to add an entry point to your Python package                                      │ https://www.youtube.com/watch?v=ERRIJUvK5Q0 │
│ if __name__ == "__main__": explained                                                  │ https://www.youtube.com/watch?v=ubpFY7_dAW4 │
│ Exploring Python dataclasses                                                          │ https://www.youtube.com/watch?v=saCd6PtvYl4 │
└───────────────────────────────────────────────────────────────────────────────────────┴─────────────────────────────────────────────┘
```

So would be nice to have a way to get the link or title+link to the clipboard without this manual cleanup step.

**Comment by bbelderbos** [comment]
@greywidget on a side note, on Mac, hovering over the links and pressing cmd, they become clickable links, very convenient. Does that work on Windows? If not the link truncating might become an issue, no?

**Comment by greywidget** [comment]
Hey Bob, I'm a bit confused by your statement "_So would be nice to have a way to get the link or title+link to the clipboard without this manual cleanup step._" Aren't they two different Issues?

- Vertical Drawing Character issue
Running `unicodedata.name() over your pipe in parenths at the top shows `VERTICAL LINE`, whereas in the pasted output it shows `BOX DRAWINGS LIGHT VERTICAL`. I suspect whether the boxes draw nicely in your terminal (or not) will be a combination of the drawing characters AND the font in your terminal.
Actually in the 3 environments I use (iTerm2 on macOS, Windows Terminal on windows, Ubuntu 20.04 on WSL2) the lines draw cleanly without showing any gaps

- Link Truncation
On all of the above the links work via `cmd+click` or `ctrl+click` but not when the link has been truncated. I'd guess that the text and link text are the same so might need to use something like wrapping

You also mention copying text the link to the textboard - if the whole text was made available is that sufficient or are you looking for an easier copying option (than select and `ctrl-c` or whatever)?

**Comment by bbelderbos** [comment]
Thanks Craig, yes they are. Let me look into this more next week, not urgent, it works good enough for me right now. How is the experience on your end?

**Comment by bbelderbos** [comment]
ah I see now, just hit it in a code review:

1. search article -> result table
2. copy one hit = │ 5 ways I use GitHub Actions                                                    │ https://pybit.es/articles/5-ways-i-use-github-actions/                                       │
3. now need to clean this up by removing | and spaces = manually effort 

Hence why I use the rust pybites-search more these days = easier copy+pasting of title+links in code reviews, compare:

```
$ pbs github acti -c a
How to Send Email Notifications Using Sendgrid and GitHub Actions
https://pybit.es/articles/how-to-send-email-notifications-using-sendgrid-and-github-actions/

5 ways I use GitHub Actions
https://pybit.es/articles/5-ways-i-use-github-actions/
...
...
```

**Comment by greywidget** [comment]
Some more context:
- The clicky link behaviour has nothing to do with `rich` or the output, it is something that is built into most terminals to recognise urls and make them clickable
- I believe that the reason partial (truncated) text works as links is down to the way that the api handles the requests. I found that `article` and `podcast` can work with partial text (I guess the podcast needs up to and including the podcast number. The others (`bites`, `tips`, `video`) will not.

I originally added the `rich` tables to try to make the ui look a bit more appealing.
Obviously I should be able to force the cells to wrap so that urls are not truncated but I noticed that this doesn't solve the _copy_ issue because (on Windows in Windows Terminal at least) selecting text for multiline (wrapped) urls or descriptions crosses boundaries between the columns so you get a bit of text from both the description AND the url.

If the copy experience can't be improved with table output (I haven't investigated this but nothing obvious springs to mind) perhaps we could add a flag to the command to specify text-only output and make sure that is in an easy-to-copy form?

**Comment by bbelderbos** [comment]
Yes, an new text-only flag would work for me and is a nice workaround for weird truncation issues.

Copy to clipboard can be left out of this discussion, when we have text only, I can just pipe it into pbcopy, but in practice at that point it's very easy to manually copy paste to a code review, so might not even need that. 

Thanks!

**Comment by greywidget** [comment]
I had a bit more of a play with this...

It's easy enough to stop the urls from truncating (they don't wrap by default as they don't contain spaces), but there are still some considerations:

- If we keep the table format and the url wraps to the next line, you can't easily copy it because text selection in the terminal is unaware of the table cell boundaries, you will get some description text and the column separator (`|`) once you start selecting on a subsequent line.
- Once the text wraps, the `ctrl+click` functionality to open in client browser breaks. The terminal doesn't know that the url continues past the line break. Note that this is true whether the url is in a `rich` table or just wrapped in the terminal (I tested on WT on Windows and ITerm2 on Mac)

I think rich tables provide a nice output format but they definitely don't work well for selecting text which wraps, as described above.

### An interesting idea to decrease the url size
Obviously, long descriptive urls are great in most cases. However the links provided by the `search` tool might be better served by something more terse. We could consider maintaining our own `short-form` links, a bit like you might see provided by products like [Bitly](https://bitly.com).
I noticed that RealPython has a tutorial on this [url shortener](https://realpython.com/build-a-python-url-shortener-with-fastapi/) which is a somewhat different use case.

So we could develop a strategy for creating/maintaining shortened links and feeding them out through the `search` tool. This would exacerbate the problem with wrapping urls (hence `ctrl+click` should work without the truncation) but if you also wanted to select the descriptive text for pasting into documentation etc that is still going to be an issue. Outputting the description and url on separate lines is probably the only way to make that play nicely.

I've attached a simple script with embedded `rich` requirement (yay, `uv`) if you want a quick play with the wrapping problems.

[main.py](https://github.com/user-attachments/files/22659096/main.py)

**Comment by bbelderbos** [comment]
Thanks Craig!

Wow that works well right? Smaller terminal:

<img width="947" height="1018" alt="Image" src="https://github.com/user-attachments/assets/7d9899a3-6d72-48ae-80eb-163f0f197fae" />

I do like the link shortener idea as well. Would we maintain that mapping locally in search or remotely on the platform where we serve the data from?

**Comment by greywidget** [comment]
It does work well visually - but there are still problems with the functionality as I mentioned above - are you seeing those issues locally (click only takes partial text (up to newline) and selecting with mouse drag pulls in Description data)?

I'd guess a link shortener would need to be handled server side. Maybe the api endpoints could talk to a common redis cache? 

Another possibility if we want to keep a ui that is a bit fancier than plain text is to use textual. We could for example have the different channels in different tabs and we wouldn't even need to display the url text - we could store it in a hidden column of a table and use buttons or shortcut keys to provide copy/browser functionality.

**Comment by bbelderbos** [comment]
I need to try/test the functionality again ... 
I like the link shortener idea but low prio for me right now. 
Then I'd rather experiment with Textual as the next iteration (oh wait, it's Hacktoberfest, right?)
Thanks.
